File: Sources/QuotaCapsuleCore/WeeklyPaceEvidence.swift
import Foundation

public enum WeeklyPaceEvidence {
    private static let day: TimeInterval = 86_400
    private static let minimumPairSeparation: TimeInterval = 30 * 60
    private static let recentHorizon: TimeInterval = 24 * 60 * 60

    public static func cycle(window: QuotaWindow, now: Date) -> PaceEvidence? {
        let duration = Double(window.windowMinutes) * 60
        let cycleStart = window.resetsAt.addingTimeInterval(-duration)
        let elapsed = now.timeIntervalSince(cycleStart)
        guard duration > 0,
              elapsed > 0,
              elapsed <= duration,
              window.resetsAt > now,
              window.usedPercent.isFinite,
              (0...100).contains(window.usedPercent) else {
            return nil
        }

        let interval = quantizedInterval(window.usedPercent)
        let elapsedDays = elapsed / day
        let elapsedFraction = min(1, max(0, elapsed / duration))
        return PaceEvidence(
            kind: .cycle,
            bandPerDay: PaceBand(
                lower: interval.lower / elapsedDays,
                upper: interval.upper / elapsedDays
            ),
            reliability: clamp(0.10 + 0.45 * sqrt(elapsedFraction), lower: 0.10, upper: 0.55),
            transitionCount: 0,
            coverageHours: elapsed / 3_600
        )
    }

    public static func recent(
        observations: [WeeklyObservation],
        now: Date
    ) -> PaceEvidence? {
        let cutoff = now.addingTimeInterval(-recentHorizon)
        let eligible = observations.filter { $0.fetchedAt >= cutoff && $0.fetchedAt <= now }
        let transitions = countUpwardTransitions(eligible)
        guard transitions > 0,
              let first = eligible.first,
              let last = eligible.last,
              let band = robustBand(eligible) else {
            return nil
        }
        let coverage = last.fetchedAt.timeIntervalSince(first.fetchedAt)
        let reliability = clamp(
            0.20
                + 0.10 * Double(min(transitions, 4))
                + 0.25 * sqrt(min(1, coverage / recentHorizon)),
            lower: 0.20,
            upper: 0.85
        )
        return PaceEvidence(
            kind: .recent,
            bandPerDay: band,
            reliability: reliability,
            transitionCount: transitions,
            coverageHours: coverage / 3_600
        )
    }

    public static func activity(
        observations: [WeeklyObservation],
        now: Date
    ) -> PaceEvidence? {
        let eligible = observations.filter { $0.fetchedAt <= now }
        let transitionIndexes = eligible.indices.dropFirst().filter {
            eligible[$0].usedPercent > eligible[$0 - 1].usedPercent
        }
        guard let firstTransition = transitionIndexes.first,
              let lastTransition = transitionIndexes.last else {
            return nil
        }

        let baseline = eligible[firstTransition - 1]
        let latestTransition = eligible[lastTransition]
        let coverage = now.timeIntervalSince(baseline.fetchedAt)
        guard coverage >= minimumPairSeparation else { return nil }

        let first = quantizedInterval(baseline.usedPercent)
        let last = quantizedInterval(latestTransition.usedPercent)
        let scale = day / coverage
        let band = PaceBand(
            lower: max(0, last.lower - first.upper) * scale,
            upper: max(0, last.upper - first.lower) * scale
        )
        guard band.upper > 0 else { return nil }

        let idle = max(0, now.timeIntervalSince(latestTransition.fetchedAt))
        let idleFactor = max(0.5, exp(-idle / (48 * 3_600)))
        let reliability = clamp(
            (0.18
                + 0.08 * Double(min(transitionIndexes.count, 4))
                + 0.18 * sqrt(min(1, coverage / recentHorizon))) * idleFactor,
            lower: 0.12,
            upper: 0.72
        )
        return PaceEvidence(
            kind: .activity,
            bandPerDay: band,
            reliability: reliability,
            transitionCount: transitionIndexes.count,
            coverageHours: coverage / 3_600
        )
    }

    public static func historical(
        observations: [WeeklyObservation],
        currentCycleID: Int
    ) -> PaceEvidence? {
        var candidates: [(band: PaceBand, reliability: Double, transitions: Int, coverage: TimeInterval)] = []
        let completedCycles = Dictionary(grouping: observations.filter { $0.cycleID != currentCycleID }, by: \.cycleID)

        for cycle in completedCycles.values {
            let segments = Dictionary(grouping: cycle, by: \.segmentID).values
            let candidate = segments.compactMap { segment -> (PaceBand, Int, TimeInterval)? in
                let ordered = segment.sorted { $0.fetchedAt < $1.fetchedAt }
                guard let first = ordered.first, let last = ordered.last else { return nil }
                let coverage = last.fetchedAt.timeIntervalSince(first.fetchedAt)
                let transitions = countUpwardTransitions(ordered)
                guard coverage >= 48 * 3_600, transitions >= 2 else { return nil }
                let start = quantizedInterval(first.usedPercent)
                let end = quantizedInterval(last.usedPercent)
                let scale = day / coverage
                return (
                    PaceBand(
                        lower: max(0, end.lower - start.upper) * scale,
                        upper: max(0, end.upper - start.lower) * scale
                    ),
                    transitions,
                    coverage
                )
            }.max { $0.2 < $1.2 }

            if let candidate {
                let reliability = clamp(
                    0.15
                        + 0.03 * Double(min(candidate.1, 4))
                        + 0.08 * min(1, candidate.2 / (7 * day)),
                    lower: 0.15,
                    upper: 0.35
                )
                candidates.append((candidate.0, reliability, candidate.1, candidate.2))
            }
        }

        guard !candidates.isEmpty else { return nil }
        return PaceEvidence(
            kind: .historical,
            bandPerDay: PaceBand(
                lower: median(candidates.map { $0.band.lower }),
                upper: median(candidates.map { $0.band.upper })
            ),
            reliability: min(0.35, candidates.map(\.reliability).reduce(0, +) / Double(candidates.count)),
            transitionCount: candidates.map(\.transitions).reduce(0, +),
            coverageHours: (candidates.map(\.coverage).max() ?? 0) / 3_600
        )
    }

    public static func fuse(_ evidence: [PaceEvidence]) -> PaceBand? {
        guard !evidence.isEmpty else { return nil }
        return PaceBand(
            lower: weightedQuantile(evidence.map { ($0.bandPerDay.lower, $0.reliability) }, quantile: 0.25),
            upper: weightedQuantile(evidence.map { ($0.bandPerDay.upper, $0.reliability) }, quantile: 0.75)
        )
    }

    public static func countUpwardTransitions(_ observations: [WeeklyObservation]) -> Int {
        zip(observations, observations.dropFirst()).reduce(into: 0) { count, pair in
            if pair.1.usedPercent > pair.0.usedPercent {
                count += 1
            }
        }
    }

    static func quantizedInterval(_ value: Double) -> PercentageBand {
        PercentageBand(
            lower: max(0, value - 0.5),
            upper: min(100, value + 0.5)
        )
    }

    private static func robustBand(_ observations: [WeeklyObservation]) -> PaceBand? {
        struct Candidate {
            let lower: Double
            let upper: Double
            var midpoint: Double { (lower + upper) / 2 }
        }

        var candidates: [Candidate] = []
        for earlierIndex in observations.indices {
            for laterIndex in observations.indices where laterIndex > earlierIndex {
                let earlier = observations[earlierIndex]
                let later = observations[laterIndex]
                let duration = later.fetchedAt.timeIntervalSince(earlier.fetchedAt)
                guard duration >= minimumPairSeparation else { continue }
                let first = quantizedInterval(earlier.usedPercent)
                let second = quantizedInterval(later.usedPercent)
                let scale = day / duration
                candidates.append(Candidate(
                    lower: max(0, second.lower - first.upper) * scale,
                    upper: max(0, second.upper - first.lower) * scale
                ))
            }
        }
        guard !candidates.isEmpty else { return nil }

        if candidates.count >= 5 {
            let center = median(candidates.map(\.midpoint))
            let mad = median(candidates.map { abs($0.midpoint - center) })
            let tolerance = max(0.01, 3 * mad)
            candidates = candidates.filter { abs($0.midpoint - center) <= tolerance }
        }
        guard !candidates.isEmpty else { return nil }
        return PaceBand(
            lower: median(candidates.map(\.lower)),
            upper: median(candidates.map(\.upper))
        )
    }

    private static func weightedQuantile(
        _ values: [(value: Double, weight: Double)],
        quantile: Double
    ) -> Double {
        let ordered = values.sorted { $0.value < $1.value }
        let total = ordered.reduce(0) { $0 + max(0.001, $1.weight) }
        var cumulative = 0.0
        for item in ordered {
            cumulative += max(0.001, item.weight)
            if cumulative >= total * quantile { return item.value }
        }
        return ordered.last?.value ?? 0
    }

    private static func median(_ values: [Double]) -> Double {
        let sorted = values.sorted()
        let middle = sorted.count / 2
        return sorted.count.isMultiple(of: 2)
            ? (sorted[middle - 1] + sorted[middle]) / 2
            : sorted[middle]
    }

    private static func clamp(_ value: Double, lower: Double, upper: Double) -> Double {
        min(upper, max(lower, value))
    }
}

File: packages/core/src/weekly-pace-evidence.ts
import type { PaceBand, PaceEvidence, PercentageBand, QuotaWindow, WeeklyObservation } from "./model";

const DAY_MS = 86_400_000;
const MINIMUM_PAIR_MS = 30 * 60_000;
const RECENT_HORIZON_MS = 24 * 60 * 60_000;
const ACTIVITY_HORIZON_MS = 72 * 60 * 60_000;
const BURST_GAP_MS = 3 * 60 * 60_000;
const ORDINARY_GAP_MS = 12 * 60 * 60_000;

export type ActivitySegmentSummary = {
  activeBurstHours: number;
  ordinaryUseHours: number;
  idleHours: number;
  dutyRatio: number;
  transitionCount: number;
  coverageHours: number;
  idleSinceLastTransitionHours: number;
  observedIncreaseBand: PaceBand;
};

export function cycleEvidence(window: QuotaWindow, now: Date): PaceEvidence | null {
  const duration = window.windowMinutes * 60_000;
  const cycleStart = window.resetsAt.getTime() - duration;
  const elapsed = now.getTime() - cycleStart;
  if (duration <= 0 || elapsed <= 0 || elapsed > duration || window.resetsAt.getTime() <= now.getTime()
    || !Number.isFinite(window.usedPercent) || window.usedPercent < 0 || window.usedPercent > 100) return null;

  const interval = quantizedInterval(window.usedPercent);
  const elapsedDays = elapsed / DAY_MS;
  const elapsedFraction = clamp(elapsed / duration, 0, 1);
  return {
    kind: "cycle",
    bandPerDay: { lower: interval.lower / elapsedDays, upper: interval.upper / elapsedDays },
    reliability: clamp(0.10 + 0.45 * Math.sqrt(elapsedFraction), 0.10, 0.55),
    transitionCount: 0,
    coverageHours: elapsed / 3_600_000,
  };
}

export function recentEvidence(observations: WeeklyObservation[], now: Date): PaceEvidence | null {
  const cutoff = now.getTime() - RECENT_HORIZON_MS;
  const eligible = observations.filter((item) => item.fetchedAt.getTime() >= cutoff && item.fetchedAt.getTime() <= now.getTime());
  const transitions = countUpwardTransitions(eligible);
  const first = eligible[0];
  const last = eligible.at(-1);
  const band = transitions > 0 ? robustBand(eligible) : null;
  if (!first || !last || !band) return null;

  const coverage = last.fetchedAt.getTime() - first.fetchedAt.getTime();
  return {
    kind: "recent",
    bandPerDay: band,
    reliability: clamp(0.20 + 0.10 * Math.min(transitions, 4) + 0.25 * Math.sqrt(Math.min(1, coverage / RECENT_HORIZON_MS)), 0.20, 0.85),
    transitionCount: transitions,
    coverageHours: coverage / 3_600_000,
  };
}

export function activityEvidence(observations: WeeklyObservation[], now: Date): PaceEvidence | null {
  const segments = activitySegments(observations, now);
  if (!segments) return null;
  const effectiveUseHours = segments.activeBurstHours + segments.ordinaryUseHours;
  if (effectiveUseHours <= 0) return null;
  const activeScale = 24 / effectiveUseHours;
  const recencyDecay = Math.exp(-segments.idleSinceLastTransitionHours / 48);
  const band = {
    lower: segments.observedIncreaseBand.lower * activeScale * segments.dutyRatio * recencyDecay,
    upper: segments.observedIncreaseBand.upper * activeScale * segments.dutyRatio * recencyDecay,
  };
  if (band.upper <= 0) return null;
  const diversity = segments.activeBurstHours > 0 && segments.ordinaryUseHours > 0 ? 0.05 : 0;
  const baseReliability = 0.18
    + 0.08 * Math.min(segments.transitionCount, 4)
    + 0.15 * Math.sqrt(Math.min(1, segments.coverageHours / 72))
    + 0.12 * Math.min(1, segments.dutyRatio * 3)
    + diversity;
  return {
    kind: "activity",
    bandPerDay: band,
    reliability: clamp(baseReliability * (0.35 + 0.65 * recencyDecay), 0.12, 0.75),
    transitionCount: segments.transitionCount,
    coverageHours: segments.coverageHours,
  };
}

export function activitySegments(observations: WeeklyObservation[], now: Date): ActivitySegmentSummary | null {
  const cutoff = now.getTime() - ACTIVITY_HORIZON_MS;
  const eligible = observations
    .filter((item) => item.fetchedAt.getTime() >= cutoff && item.fetchedAt.getTime() <= now.getTime())
    .slice()
    .sort((left, right) => left.fetchedAt.getTime() - right.fetchedAt.getTime());
  const first = eligible[0];
  const last = eligible.at(-1);
  if (!first || !last || eligible.length < 2) return null;
  const coverage = now.getTime() - first.fetchedAt.getTime();
  if (coverage < MINIMUM_PAIR_MS) return null;

  let active = 0;
  let ordinary = 0;
  let idle = 0;
  let transitions = 0;
  let observedIncreaseLower = 0;
  let observedIncreaseUpper = 0;
  let lastTransitionAt: Date | null = null;
  let measurementStart = first;
  let measurementEnd = first;
  const flushMeasurement = () => {
    if (measurementEnd.usedPercent <= measurementStart.usedPercent) return;
    const startInterval = quantizedInterval(measurementStart.usedPercent);
    const endInterval = quantizedInterval(measurementEnd.usedPercent);
    observedIncreaseLower += Math.max(0, endInterval.lower - startInterval.upper);
    observedIncreaseUpper += Math.max(0, endInterval.upper - startInterval.lower);
  };
  for (let index = 1; index < eligible.length; index += 1) {
    const earlier = eligible[index - 1];
    const later = eligible[index];
    const gap = later.fetchedAt.getTime() - earlier.fetchedAt.getTime();
    if (gap <= 0) continue;
    if (later.usedPercent > earlier.usedPercent) {
      transitions += 1;
      lastTransitionAt = later.fetchedAt;
      if (gap <= BURST_GAP_MS) active += gap;
      else if (gap <= ORDINARY_GAP_MS) ordinary += gap;
      else {
        ordinary += BURST_GAP_MS;
        idle += gap - BURST_GAP_MS;
      }
    } else {
      idle += gap;
    }
    if (later.usedPercent < earlier.usedPercent) {
      flushMeasurement();
      measurementStart = later;
      measurementEnd = later;
    } else {
      measurementEnd = later;
    }
  }
  flushMeasurement();
  idle += Math.max(0, now.getTime() - last.fetchedAt.getTime());
  if (!transitions || observedIncreaseUpper <= 0 || !lastTransitionAt) return null;
  return {
    activeBurstHours: active / 3_600_000,
    ordinaryUseHours: ordinary / 3_600_000,
    idleHours: idle / 3_600_000,
    dutyRatio: clamp((active + ordinary) / coverage, 0, 1),
    transitionCount: transitions,
    coverageHours: coverage / 3_600_000,
    idleSinceLastTransitionHours: Math.max(0, now.getTime() - lastTransitionAt.getTime()) / 3_600_000,
    observedIncreaseBand: {
      lower: observedIncreaseLower,
      upper: observedIncreaseUpper,
    },
  };
}

export function historicalEvidence(observations: WeeklyObservation[], currentCycleID: number): PaceEvidence | null {
  const cycles = new Map<number, WeeklyObservation[]>();
  for (const observation of observations) {
    if (observation.cycleID === currentCycleID) continue;
    const cycle = cycles.get(observation.cycleID) ?? [];
    cycle.push(observation);
    cycles.set(observation.cycleID, cycle);
  }

  const candidates: Array<{ band: PaceBand; reliability: number; transitions: number; coverage: number }> = [];
  for (const cycle of cycles.values()) {
    const segments = new Map<number, WeeklyObservation[]>();
    for (const observation of cycle) {
      const segment = segments.get(observation.segmentID) ?? [];
      segment.push(observation);
      segments.set(observation.segmentID, segment);
    }
    const segmentCandidates = [...segments.values()].map((segment) => {
      const ordered = segment.slice().sort((left, right) => left.fetchedAt.getTime() - right.fetchedAt.getTime());
      const first = ordered[0];
      const last = ordered.at(-1);
      if (!first || !last) return null;
      const coverage = last.fetchedAt.getTime() - first.fetchedAt.getTime();
      const transitions = countUpwardTransitions(ordered);
      if (coverage < 48 * 3_600_000 || transitions < 2) return null;
      const start = quantizedInterval(first.usedPercent);
      const end = quantizedInterval(last.usedPercent);
      const scale = DAY_MS / coverage;
      return {
        band: { lower: Math.max(0, end.lower - start.upper) * scale, upper: Math.max(0, end.upper - start.lower) * scale },
        reliability: clamp(0.15 + 0.03 * Math.min(transitions, 4) + 0.08 * Math.min(1, coverage / (7 * DAY_MS)), 0.15, 0.35),
        transitions,
        coverage,
      };
    }).filter((item): item is NonNullable<typeof item> => item !== null)
      .sort((left, right) => right.coverage - left.coverage);
    if (segmentCandidates[0]) candidates.push(segmentCandidates[0]);
  }

  if (!candidates.length) return null;
  return {
    kind: "historical",
    bandPerDay: {
      lower: median(candidates.map((item) => item.band.lower)),
      upper: median(candidates.map((item) => item.band.upper)),
    },
    reliability: Math.min(0.35, candidates.reduce((sum, item) => sum + item.reliability, 0) / candidates.length),
    transitionCount: candidates.reduce((sum, item) => sum + item.transitions, 0),
    coverageHours: Math.max(...candidates.map((item) => item.coverage)) / 3_600_000,
  };
}

export function fusePaceEvidence(evidence: PaceEvidence[]): PaceBand | null {
  if (!evidence.length) return null;
  return {
    lower: weightedQuantile(evidence.map((item) => [item.bandPerDay.lower, item.reliability]), 0.25),
    upper: weightedQuantile(evidence.map((item) => [item.bandPerDay.upper, item.reliability]), 0.75),
  };
}

export function countUpwardTransitions(observations: WeeklyObservation[]): number {
  let count = 0;
  for (let index = 1; index < observations.length; index += 1) {
    if (observations[index].usedPercent > observations[index - 1].usedPercent) count += 1;
  }
  return count;
}

export function quantizedInterval(value: number): PercentageBand {
  return { lower: Math.max(0, value - 0.5), upper: Math.min(100, value + 0.5) };
}

function robustBand(observations: WeeklyObservation[]): PaceBand | null {
  let candidates: Array<PaceBand & { midpoint: number }> = [];
  for (let earlier = 0; earlier < observations.length; earlier += 1) {
    for (let later = earlier + 1; later < observations.length; later += 1) {
      const duration = observations[later].fetchedAt.getTime() - observations[earlier].fetchedAt.getTime();
      if (duration < MINIMUM_PAIR_MS) continue;
      const first = quantizedInterval(observations[earlier].usedPercent);
      const second = quantizedInterval(observations[later].usedPercent);
      const scale = DAY_MS / duration;
      const lower = Math.max(0, second.lower - first.upper) * scale;
      const upper = Math.max(0, second.upper - first.lower) * scale;
      candidates.push({ lower, upper, midpoint: (lower + upper) / 2 });
    }
  }
  if (!candidates.length) return null;
  if (candidates.length >= 5) {
    const center = median(candidates.map((item) => item.midpoint));
    const mad = median(candidates.map((item) => Math.abs(item.midpoint - center)));
    const tolerance = Math.max(0.01, 3 * mad);
    candidates = candidates.filter((item) => Math.abs(item.midpoint - center) <= tolerance);
  }
  return candidates.length ? {
    lower: median(candidates.map((item) => item.lower)),
    upper: median(candidates.map((item) => item.upper)),
  } : null;
}

function weightedQuantile(values: Array<[number, number]>, quantile: number): number {
  const ordered = values.slice().sort((left, right) => left[0] - right[0]);
  const total = ordered.reduce((sum, item) => sum + Math.max(0.001, item[1]), 0);
  let cumulative = 0;
  for (const item of ordered) {
    cumulative += Math.max(0.001, item[1]);
    if (cumulative >= total * quantile) return item[0];
  }
  return ordered.at(-1)?.[0] ?? 0;
}

function median(values: number[]): number {
  const ordered = values.slice().sort((left, right) => left - right);
  const middle = Math.floor(ordered.length / 2);
  return ordered.length % 2 === 0 ? (ordered[middle - 1] + ordered[middle]) / 2 : ordered[middle];
}

function clamp(value: number, lower: number, upper: number): number {
  return Math.min(upper, Math.max(lower, value));
}
